src/common/io-providers.ts

```diff
@@ -0,0 +1,1246 @@
+/**
+ * Copyright (C) 2022-2024 Permanent Data Solutions, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  AoPrimaryName,
+  AoPrimaryNameRequest,
+  WalletAddress,
+} from '../types/index.js';
+import { AOProcess } from './contracts/ao-process.js';
+import { HB } from './hyperbeam/hb.js';
+import { Logger } from './logger.js';
+
+/**
+ * Base interface for ARIO read providers
+ */
+export interface ARIOReadProvider {
+  getBalance(params: { address: WalletAddress }): Promise<number>;
+  getPrimaryNameRequest(params: {
+    initiator: WalletAddress;
+  }): Promise<AoPrimaryNameRequest>;
+  getPrimaryName(
+    params: { address: WalletAddress } | { name: string },
+  ): Promise<AoPrimaryName>;
+}
```

Comment:
suggestion: use the existing AoARIORead interface and implement that. for any functions not yet supported throw an Error